shared/constants/metametrics.js

```diff
@@ -0,0 +1,213 @@
+// Type Imports
+/**
+ * @typedef {import('../../app/scripts/lib/enums').EnvironmentType} EnvironmentType
+ */
+
+// Type Declarations
+/**
+ * Used to attach context of where the user was at in the application when the
+ * event was triggered. Also included as full details of the current page in
+ * page events.
+ * @typedef {Object} MetaMetricsPageObject
+ * @property {string} [path] - the path of the current page (e.g /home)
+ * @property {string} [title] - the title of the current page (e.g 'home')
+ * @property {string} [url] - the fully qualified url of the current page
+ */
+
+/**
+ * For metamask, this is the dapp that triggered an interaction
+ * @typedef {Object} MetaMetricsReferrerObject
+ * @property {string} [url] - the origin of the dapp issuing the
+ *  notification
+ */
+
+/**
+ * We attach context to every meta metrics event that help to qualify our
+ * analytics. This type has all optional values because it represents a
+ * returned object from a method call. Ideally app and userAgent are
+ * defined on every event. This is confirmed in the getTrackMetaMetricsEvent
+ * function, but still provides the consumer a way to override these values if
+ * necessary.
+ * @typedef {Object} MetaMetricsContext
+ * @property {Object} app
+ * @property {string} app.name - the name of the application tracking the event
+ * @property {string} app.version - the version of the application
+ * @property {string} userAgent - the useragent string of the user
+ * @property {MetaMetricsPageObject} [page] - an object representing details of
+ *  the current page
+ * @property {MetaMetricsReferrerObject} [referrer] - for metamask, this is the
+ *  dapp that triggered an interaction
+ */
+
+/**
+ * @typedef {Object} MetaMetricsEventPayload
+ * @property {string}  event - event name to track
+ * @property {string}  category - category to associate event to
+ * @property {string} [environmentType] - The type of environment this event
+ *  occurred in. Defaults to the background process type
+ * @property {object}  [properties] - object of custom values to track, keys
+ *  in this object must be in snake_case
+ * @property {object}  [sensitiveProperties] - Object of sensitive values to
+ *  track. Keys in this object must be in snake_case. These properties will be
+ *  sent in an additional event that excludes the user's metaMetricsId
+ * @property {number}  [revenue] - amount of currency that event creates in
+ *  revenue for MetaMask
+ * @property {string}  [currency] - ISO 4127 format currency for events with
+ *  revenue, defaults to US dollars
+ * @property {number}  [value] - Abstract business "value" attributable to
+ *  customers who trigger this event
+ * @property {MetaMetricsDynamicContext} [eventContext] - additional context
```

Comment:
Is there still a `typedef` for `MetaMetricsDynamicContext` somewhere? I saw one deleted in `shared/modules/metametrics.js` but I can't find one added.

Reply:
Nope, this property no longer exists and instead [page] and [referrer] are top-level properties that apparently never got updated. I really want to figure out how to make typescript check our JSdocs only and like, ignore all of the other untyped code <_< (until we get full support obviously)

Reply:
resolved in https://github.com/MetaMask/metamask-extension/pull/9857/commits/673a6508e7dc7c53cb04e6cd38829da33aae15d1